boardesp/proxy.c

```diff
@@ -51,10 +51,14 @@ static int ICACHE_FLASH_ATTR __spi_comm(char *dat, int len, uint32_t *recvData,
   spiData.dataLen = 0x14;
   SPIMasterSendData(SpiNum_HSPI, &spiData);
 
-  // give the ST time to be ready, 1 ms
-  // TODO: how can we do this better?
-  os_delay_us(1000);
-
+  // use the boot0 pin as a handshake with the ST
+  // set boot0 as input
+  gpio_output_set(0,0,0,(1 << 4));
+  //wait for boot0 to be set to LOW by ST, counter to avoid infinite loop
+  for(int i = 0;(gpio_input_get() & (1 << 4)) && i < 100; i++) {
```

Comment:
The counter makes it wait for 1ms, same as before. If the data isn't ready it will be an issue. The only other way I can think of would be to use an infinite loop and always poll but that would crash the system if spi fails. Even using a ack would require a counter to avoid an infinite loop. Any suggestions?